jax/_src/scipy/linalg.py

```diff
@@ -2248,3 +2248,155 @@ def _binom(n, k):
+    method: "schur" is the default and is accurate but slow, and "eigen" is an alternative that is faster but less accurate for ill-conditioned matrices.
+    tol: How close the sum of the eigenvalues from A and B can be to zero before returning matrix of NaNs
+
+  Returns:
+    X: Matrix of shape m x n
+
+  Examples:
+    >>> A = jax.numpy.array([[1, 2], [3, 4]])
+    >>> B = jax.numpy.array([[5, 6], [7, 8]])
+    >>> C = jax.numpy.array([[6, 8], [10, 12]])
+    >>> X = jax.scipy.linalg.solve_sylvester(A, B, C)
+    >>> with jax.numpy.printoptions(precision=0):
+    ...   print(X)
+    [[ 1.e+00 -4.e-07]
+      [-3.e-07  1.e+00]]
+
+  Notes:
+    The Bartel-Stewart algorithm is robust because a Schur decomposition always exists even for defective matrices,
+    and it handles complex and ill-conditioned problems better than the eigen decomposition method.
+    However, there are a couple of drawbacks. First, It is computationally more expensive than
+    the eigen decomposition method because you need to perform a Schur decomposition and then scan the entire solution matrix.
+    Second, it requires more system memory compared to the eigen decomposition method.
+
+    The eigen decomposition method is the fastest method to solve a sylvester equation. However, this speed brings with it a couple of drawbacks.
+    First, A and B must be diagonalizable otherwise the eigenvectors will be linearly dependent and ill-conditioned leading to accuracy issues.
+    Second, when the eigenvectors are not orthogonal roundoff errors are amplified.
+
+    The tol argument allows you to specify how ill-conditioned a matrix can be and still estimate a solution.
+    For matrices that are ill-conditioned we recommend using float64 instead of the default float32 dtype. The solver
+    can still return good estimates for ill-conditioned matrices depending on how close to zero the sums of the eigenvalues of A and B
+    are.
+  """
+  A, B, C = promote_args_inexact("solve_sylvester", A, B, C)
+
+  m, n = C.shape
+
+  if A.shape != (m, m) or B.shape != (n, n) or C.shape != (m, n):
+    raise ValueError(f"Incompatible shapes for Sylvester equation:\nA: {A.shape}\nB: {B.shape}\nC: {C.shape}")
+
+  if method == "schur":
+    # Schur decomposition
+    R, U = schur(A, output='complex')
+    S, V = schur(B.conj().T, output='complex')
+
+    # Transform right-hand side
+    F = U.conj().T @ C.astype(R.dtype) @ V
+
+    # Solve triangular Sylvester system
+    Y = _solve_sylvester_triangular_scan(R, S.conj().T, F)
+
+    # Transform back
+    X = U @ Y @ V.conj().T
+  elif method == "eigen":
+    RA, UA = jnp.linalg.eig(A)
+    RB, UB = jnp.linalg.eig(B)
+    F = solve(UA, C.astype(RA.dtype) @ UB)
+    W = RA[:, None] + RB[None, :]
+    Y = F / W
+    X = UA[:m,:m] @ Y[:m,:n] @ inv(UB)[:n,:n]
+  else:
+    raise ValueError(f"Unrecognized method {method}. The two valid methods are either \"schur\" or \"eigen\".")
+
+  X = jnp.real(X) if jnp.isrealobj(C) else X
```

Comment:
`jnp.isrealobj` is not available here after the recent build refactor. I'd rewrite it like this:
```python
if dtypes.issubdtype(C.dtype, np.complexfloating):
  X = X.real
```